src/parsec_service/stability.md

```diff
@@ -0,0 +1,68 @@
+# Parsec Stability
+
+Work in progress, see the [related issue](https://github.com/parallaxsecond/parsec-book/issues/83).
+
+## Why is stability important?
+
+Parsec is a long-running service that will be configured, deployed, upgraded and maintained over a
+long period of time. The environment will also have client applications that are calling it, and
+persistent state that is collected on storage. Components will be upgraded over time. For Parsec to
+be supportable in enterprise system, it must not be brittle against upgrades. If Parsec is upgraded
+and restarted, client applications should continue working, and all persistent state from prior
+versions should remain valid and usable.
+
+## What does stability mean for the Parsec service?
+
+The following definition of stability is chosen for the Parsec service: *whatever two versions A and
+B of the Parsec binary, B being newer than A, A and B are stable if A can be removed and replaced by
+B without breaking anything*.
+
+Note that we are not looking at stability in the reverse direction: B can not be replaced by A
+without breaking anything.
+
+The principle of [semantic versioning](https://semver.org/) is used to describe the stability of
+Parsec versions and when breaking changes are done. If `parsec` version is at `1.0.0` then all
+future stable version to that will be `1.x.y`.
+
+## What needs to be stable?
+
+With the above definition, Parsec stays stable if it keeps the exact same communication channels and
+format with its environment. Those are already described in the [dataflow
+diagram](../parsec_security/parsec_threat_model/threat_model.md#dataflow-diagram) in the threat
+model.
+
+The communication channels that needs to be stable are (some of them are not in the diagram):
+
+1. Communication with clients
+2. Communication with authenticators
+3. Communication received from the CLI invocation
+4. Configuration file (including default configuration options)
+5. Key mappings
+6. OS signals
+7. systemd communication
+8. Dynamic libraries dependencies
```

Comment:
I think there's another thing needed here: 
```
9. Provider workflow and call mapping
```
The name isn't the best, sorry. Imagine a key being created in the PKCS11 provider, restricted to some mechanism based on the key attributes sent in `PsaGenerateKey`. Then in the next version of the service, we make some changes - even with the same key attributes, we now create the key without a restriction on the mechanism, and instead specify it in some other way (when we make the call for signing, for example). Maybe this isn't allowed if the key already has those properties set by the previous version of the service, and the operation now fails. I don't know if the example works with PKCS11, it sounds vaguely familiar, but it's definitely a thing for TPMs.

I think this is tricky to define, especially because the number of options is huge. Maybe we can say that
1) objects created by the providers in their backends should be immutable
2) the way those immutable objects are created should not change between stable versions
???

Reply:
I see what you mean and for me this is part of the contracts' definition in the wire protocol (so part of point 1).
By that I mean that if [PsaGenerateKey](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/psa_generate_key.html) is modified so that the policies do not work in the same way, then there would need to be a new GenerateKey operation in the new behaviour.
For me the contract both describes the format of the operation (protobuf) and also its behaviour. Those two things should not break I think.

Reply:
It's not strictly about that, because you could reasonably implement the same operation that behaves the same way and correctly follows the Parsec frontend API, but uses the backend API in different ways, especially when it comes to the exact list of attributes we assign to the key. And this varies quite a lot from one backend to another.

Think about the checks for whether a key can be used to sign, within the TPM provider. In version 1.0.0 the checks might be done in software in the provider. If we then move, in version 1.1.0, to add the attributes on the key so that the TPM itself returns an error if someone tries to sign with a key that's not intended for signing, we can remove the software checks. All is fine, because both versions correctly implement the same interface and honor the contract in the same way. But a key created in 1.0.0 will not necessarily result in an error if it is misused, because we didn't set the correct attributes.

Reply:
Thanks for the precision and the example.

I still thinks it's related with respecting the contract (so point 1): no matter at what version a key is created (and then used), it should respect the contract as described in the key generation operation. If we introduce a bug where the contract is not longer respected when a key created in 1.0.0 is used in 1.1.0 then the stability of the communication channel with clients has been broken.

That is a good point though I did not think about. I can add a sentence about operation contract being stable in point 1. Testing that might need some new things that we currently do not have. For keys for example, their policies should be respected at at Parsec version.

Reply:
I think it's a bit of a stretch to include it in 1 - the same could be said about `5. Key mappings`, that no matter at what version a key mapping is created, it should allow providers to respect their contracts. My example was restricted to a change in behaviour, but it could very well be that some changes in the way we use the backend leaves all old keys completely unusable.

It would indeed be very nice if we could test, and I'm sure we can. For all of them we have the keys and all the configuration stored on disk, there should be a way to preserve an environment with a bunch of keys provisioned and test in it. This would also cover the testing for key mappings. I'll raise an issue in the service repo

Reply:
I agree that 1 then becomes big but I still think that the respect of the contract is included in the communication channel with the client. The contract is what the client assumes Parsec will respect. I can split 1 in 3 parts to make it more readable/easier:
- definition of the format of requests/responses (ie the format)
- definition of the operation contracts (ie the logic)
- definition of the listener endpoints

About testing in general for this document, my plan was, once it is agreed, to create one single specific issue for each line of the table where we will look at how to stabilise and how to test.